## Bahasa Indonesia
Sebelum kita melakukan manipulasi data, ada beberapa library yang harus kita import terlebih dahulu, yakni:
1. pandas untuk membaca data di python menggunakan format dataframe
2. psycopg2 sebagai PostgreSQL database adapter
3. sqlalchemy sebagai Toolkit dan Object Relational Mapper SQL
Selain itu kita juga perlu meng-import create_engine dari sqlalchemy sebagai connector.

## English
Before we manipulate the data, there are several libraries that we should import first, they are:
1. pandas to read data in Python with Dataframe format
2. psycopg2 as a PostgreSQL database adapter
3. sqlalchemy as a Toolkit and Object Relational Mapper SQL
Besides that we also need to import create_engine from sqlalemy as a connector.

In [1]:
import pandas as pd
import psycopg2
import sqlalchemy
from sqlalchemy import create_engine
%matplotlib inline

In [2]:
#call the data from local database
postgres_connection = ('postgresql://postgres:iykra@localhost:5432/postgres')
cnx = create_engine(postgres_connection)
#make a new dataframe from the database
df = pd.read_sql_query("""select*from customer""",cnx)
#print the dataframe
df.head()

customer_id         country
0        12346  United Kingdom
1        12347         Iceland
2        12348         Finland
3        12349           Italy
4        12350          Norway

In [3]:
#create day of week (dow) name on transaction
df_dow_name = pd.read_sql_query("""
select *,
case
when extract(dow from invoice_date) = 0 then 'Sunday'
when extract(dow from invoice_date) = 1 then 'Monday'
when extract(dow from invoice_date) = 2 then 'Tuesday'
when extract(dow from invoice_date) = 3 then 'Wednesday'
when extract(dow from invoice_date) = 4 then 'Thursday'
when extract(dow from invoice_date) = 5 then 'Friday'
when extract(dow from invoice_date) = 6 then 'Saturday'
end as day_of_week_name
from transaction
""",cnx)
df_dow_name.head()

trx_id invoice_no stock_code  quantity        invoice_date  unit_price  \
0       1     536365     85123A         6 2010-12-01 08:26:00        2.55   
1       2     536365      71053         6 2010-12-01 08:26:00        3.39   
2       3     536365     84406B         8 2010-12-01 08:26:00        2.75   
3       4     536365     84029G         6 2010-12-01 08:26:00        3.39   
4       5     536365     84029E         6 2010-12-01 08:26:00        3.39   

   customer_id day_of_week_name  
0      17850.0        Wednesday  
1      17850.0        Wednesday  
2      17850.0        Wednesday  
3      17850.0        Wednesday  
4      17850.0        Wednesday

In [4]:
#create weekend or weekday categories on transaction
df_weekday = pd.read_sql_query("""select *,
case when extract(dow from invoice_date) in (0,6) then 'weekend' 
else 'weekday' 
end as day_of_week
from transaction
where invoice_no='536988'
""",cnx)
df_weekday.head()

trx_id invoice_no stock_code  quantity        invoice_date  unit_price  \
0    7075     536988      22086         2 2010-12-03 15:05:00        2.95   
1    7076     536988      21918         1 2010-12-03 15:05:00        0.42   
2    7077     536988     47343A         3 2010-12-03 15:05:00        1.49   
3    7078     536988      22142         2 2010-12-03 15:05:00        1.45   
4    7079     536988      21098         1 2010-12-03 15:05:00        1.25   

   customer_id day_of_week  
0        17238     weekday  
1        17238     weekday  
2        17238     weekday  
3        17238     weekday  
4        17238     weekday

In [5]:
#create time of the day on transaction
df_timeofday = pd.read_sql_query("""select *,
case 
    when extract(hour from invoice_date) between 0 and 10 then 'morning'
    when extract(hour from invoice_date) between 11 and 18 then 'afternoon'
    when extract(hour from invoice_date) between 18 and 24 then 'night'
end as timeofday
from transaction""",cnx)
df_timeofday.head()

trx_id invoice_no stock_code  quantity        invoice_date  unit_price  \
0       1     536365     85123A         6 2010-12-01 08:26:00        2.55   
1       2     536365      71053         6 2010-12-01 08:26:00        3.39   
2       3     536365     84406B         8 2010-12-01 08:26:00        2.75   
3       4     536365     84029G         6 2010-12-01 08:26:00        3.39   
4       5     536365     84029E         6 2010-12-01 08:26:00        3.39   

   customer_id timeofday  
0      17850.0   morning  
1      17850.0   morning  
2      17850.0   morning  
3      17850.0   morning  
4      17850.0   morning

In [6]:
#create day of week name, weekend/weekday categories, and time of the day in a dataframe
df_combo = pd.read_sql_query("""
select *,
case
when extract(dow from invoice_date) = 0 then 'Sunday'
when extract(dow from invoice_date) = 1 then 'Monday'
when extract(dow from invoice_date) = 2 then 'Tuesday'
when extract(dow from invoice_date) = 3 then 'Wednesday'
when extract(dow from invoice_date) = 4 then 'Thursday'
when extract(dow from invoice_date) = 5 then 'Friday'
when extract(dow from invoice_date) = 6 then 'Saturday'
end as day_of_week_name,

case 
when extract(dow from invoice_date) in (0,6) then 'weekend' 
else 'weekday' 
end as day_of_week,

case 
when extract(hour from invoice_date) between 0 and 10 then 'morning'
when extract(hour from invoice_date) between 11 and 18 then 'afternoon'
when extract(hour from invoice_date) between 18 and 24 then 'night'
end as timeofday

from transaction
order by trx_id desc
""",cnx)
df_combo.head()

trx_id invoice_no stock_code  quantity        invoice_date  unit_price  \
0  541909     581587      22138         3 2011-12-09 12:50:00        4.95   
1  541908     581587      23255         4 2011-12-09 12:50:00        4.15   
2  541907     581587      23254         4 2011-12-09 12:50:00        4.15   
3  541906     581587      22899         6 2011-12-09 12:50:00        2.10   
4  541905     581587      22613        12 2011-12-09 12:50:00        0.85   

   customer_id day_of_week_name day_of_week  timeofday  
0      12680.0           Friday     weekday  afternoon  
1      12680.0           Friday     weekday  afternoon  
2      12680.0           Friday     weekday  afternoon  
3      12680.0           Friday     weekday  afternoon  
4      12680.0           Friday     weekday  afternoon

In [7]:
#analyze which combination between day of week and time of the day which has the largest value
df_analysis = pd.read_sql_query("""
with trx_time as
(
select *,
case
when extract(dow from invoice_date) = 0 then 'Sunday'
when extract(dow from invoice_date) = 1 then 'Monday'
when extract(dow from invoice_date) = 2 then 'Tuesday'
when extract(dow from invoice_date) = 3 then 'Wednesday'
when extract(dow from invoice_date) = 4 then 'Thursday'
when extract(dow from invoice_date) = 5 then 'Friday'
when extract(dow from invoice_date) = 6 then 'Saturday'
end as day_of_week_name,

case 
when extract(dow from invoice_date) in (0,6) then 'weekend' 
else 'weekday' 
end as day_of_week,

case 
when extract(hour from invoice_date) between 0 and 10 then 'morning'
when extract(hour from invoice_date) between 11 and 18 then 'afternoon'
when extract(hour from invoice_date) between 18 and 24 then 'night'
end as timeofday

from transaction)

select day_of_week_name, timeofday, count(distinct invoice_no) as total from trx_time
group by day_of_week_name, timeofday 
order by total desc
""",cnx)
df_analysis.head()

day_of_week_name  timeofday  total
0         Thursday  afternoon   4272
1        Wednesday  afternoon   3779
2          Tuesday  afternoon   3636
3           Monday  afternoon   3296
4           Friday  afternoon   3086

From the simple analysis above, we get an insight that transaction often happens on Thursday and Wednesday afternoon. From the insight, we can assume that a lot of customer doing transaction in the middle of the week.

In [8]:
#make customer transaction recency
df_recency = pd.read_sql_query("""
with customer_trx as 
(select customer_id, date(invoice_date) invoice_date
, count(distinct(invoice_no)) total_order
, count(distinct(date(invoice_date))) total_date
from transaction
where customer_id is not null
group by 1,2
order by 1,2)
, 
customer_day_diff as 
(select customer_id, invoice_date, total_order, total_date,
date(invoice_date)-date(lag(invoice_date,1) over (partition by customer_id 
order by invoice_date)) as day_diff
from customer_trx)

, customer_day_diff2 as (select customer_id, sum(total_order) total_order, sum(total_date) total_day,
avg(day_diff) avg_recency, max(day_diff) max_recency, min(day_diff) min_recency
from customer_day_diff
where day_diff is not null
group by 1
order by 1)
select *,case
when avg_recency < (SELECT PERCENTILE_CONT(0.5) WITHIN
GROUP(ORDER BY avg_recency) FROM customer_day_diff2) then 'low' else 'high' 
end as frequency_rec
from customer_day_diff2
""",cnx)
df_recency.head()

customer_id  total_order  total_day  avg_recency  max_recency  min_recency  \
0        12347          6.0        6.0    60.833333           90           37   
1        12348          3.0        3.0    94.333333          173           40   
2        12352         10.0        6.0    43.333333          182            5   
3        12356          2.0        2.0   151.500000          223           80   
4        12358          1.0        1.0   149.000000          149          149   

  frequency_rec  
0          high  
1          high  
2           low  
3          high  
4          high

In [9]:
df_recency.describe()

customer_id  total_order    total_day  avg_recency  max_recency  \
count   2991.000000  2991.000000  2991.000000  2991.000000  2991.000000   
mean   15269.896356     5.830491     4.989636    67.018067   103.691408   
std     1721.090569    10.721888     7.566708    63.677906    74.390123   
min    12347.000000     1.000000     1.000000     1.000000     1.000000   
25%    13795.000000     1.000000     1.000000    25.309524    48.000000   
50%    15221.000000     3.000000     3.000000    47.666667    91.000000   
75%    16770.500000     6.000000     6.000000    85.000000   142.000000   
max    18287.000000   246.000000   145.000000   366.000000   366.000000   

       min_recency  
count  2991.000000  
mean     41.762287  
std      67.634884  
min       1.000000  
25%       4.000000  
50%      12.000000  
75%      44.000000  
max     366.000000

In [10]:
#How Big Basket Size
df_basketsize = pd.read_sql_query("""
select customer_id,sum(unit_price * quantity) as pricetotal,
count(distinct invoice_no) as Jumlah_trx,
sum(unit_price * quantity)/ count(distinct invoice_no) as basketsize,
max(unit_price * quantity)/count(distinct invoice_no) as max_basketsize,
min(unit_price * quantity)/count(distinct invoice_no) as min_basketsize,
avg(unit_price * quantity)/ count(distinct invoice_no) as avg
from transaction
where customer_id is not null and quantity>0
group by customer_id order by basketsize desc
""",cnx)
df_basketsize.head()

customer_id  pricetotal  jumlah_trx    basketsize  max_basketsize  \
0        16446   168472.50           2  84236.250000         84234.8   
1        12346    77183.60           1  77183.600000         77183.6   
2        15749    44534.30           3  14844.766667          2179.8   
3        15098    39916.50           3  13305.500000         12990.0   
4        12357     6207.67           1   6207.670000           237.6   

   min_basketsize           avg  
0           0.625  28078.750000  
1       77183.600  77183.600000  
2         910.000   1484.476667  
3          99.000   4435.166667  
4          10.080     47.386794

In [11]:
#customer transactin frequency
df_frequency = pd.read_sql_query("""
with customer_frequency as 
(select customer_id, 
extract(year from invoice_date) year_number, 
extract(month from invoice_date) month_number, 
count(distinct(invoice_no)) total_order
from transaction
where customer_id is not null
group by 1,2,3
order by 1,2,3)

select customer_id, avg(total_order) avg_total_order_monthly, 
max(total_order) max_total_order, min(total_order) min_total_order
from customer_frequency
group by 1
""",cnx)
df_frequency.head()

customer_id  avg_total_order_monthly  max_total_order  min_total_order
0        16592                 2.375000                4                1
1        14173                 1.333333                2                1
2        13527                 1.600000                3                1
3        14067                 1.000000                1                1
4        12502                 1.000000                1                1

In [12]:
freq_rec =  pd.merge(df_recency,df_frequency,how='inner')
freq_rec.head()

customer_id  total_order  total_day  avg_recency  max_recency  min_recency  \
0        12347          6.0        6.0    60.833333           90           37   
1        12348          3.0        3.0    94.333333          173           40   
2        12352         10.0        6.0    43.333333          182            5   
3        12356          2.0        2.0   151.500000          223           80   
4        12358          1.0        1.0   149.000000          149          149   

  frequency_rec  avg_total_order_monthly  max_total_order  min_total_order  
0          high                     1.00                1                1  
1          high                     1.00                1                1  
2           low                     2.75                7                1  
3          high                     1.00                1                1  
4          high                     1.00                1                1

In [13]:
df_monetery = pd.read_sql_query("""
with
customer_monitize as (select customer_id,sum(unit_price * quantity) as pricetotal,
count(distinct invoice_no) as Jumlah,
sum(unit_price * quantity)/ count(distinct invoice_no) as basketsize,
max(unit_price * quantity)/count(distinct invoice_no) as max,
min(unit_price * quantity)/count(distinct invoice_no) as min,
avg(unit_price * quantity)/ count(distinct invoice_no) as avg
from transaction
where customer_id is not null and quantity>0
group by customer_id order by basketsize desc)
select *,case
when avg < (SELECT PERCENTILE_CONT(0.5) WITHIN
GROUP(ORDER BY avg) FROM customer_monitize) then 'low' else 'high'
end as frequency_mrkt
from customer_monitize;
""",cnx)
df_monetery.head()

customer_id  pricetotal  jumlah    basketsize      max        min  \
0        16446   168472.50       2  84236.250000  84234.8      0.625   
1        12346    77183.60       1  77183.600000  77183.6  77183.600   
2        15749    44534.30       3  14844.766667   2179.8    910.000   
3        15098    39916.50       3  13305.500000  12990.0     99.000   
4        12357     6207.67       1   6207.670000    237.6     10.080   

            avg frequency_mrkt  
0  28078.750000           high  
1  77183.600000           high  
2   1484.476667           high  
3   4435.166667           high  
4     47.386794           high

In [14]:
frm = pd.merge(freq_rec,df_monetery, how='inner')
frm.head()

customer_id  total_order  total_day  avg_recency  max_recency  min_recency  \
0        12347          6.0        6.0    60.833333           90           37   
1        12348          3.0        3.0    94.333333          173           40   
2        12352         10.0        6.0    43.333333          182            5   
3        12356          2.0        2.0   151.500000          223           80   
4        12358          1.0        1.0   149.000000          149          149   

  frequency_rec  avg_total_order_monthly  max_total_order  min_total_order  \
0          high                     1.00                1                1   
1          high                     1.00                1                1   
2           low                     2.75                7                1   
3          high                     1.00                1                1   
4          high                     1.00                1                1   

   pricetotal  jumlah  basketsize         max     min        avg  \
0     4310.00       7  615.714286   35.657143  0.7200   3.383046   
1     1797.24       4  449.310000   60.000000  3.3000  14.493871   
2     2506.04       8  313.255000   47.062500  1.2375   3.685353   
3     2811.43       3  937.143333  108.000000  1.2500  15.883785   
4     1168.06       2  584.030000   80.000000  4.6800  30.738421   

  frequency_mrkt  
0            low  
1           high  
2            low  
3           high  
4           high

In [15]:
customer = pd.read_sql_query("""
select * from customer""",cnx)
customer.head()

customer_id         country
0        12346  United Kingdom
1        12347         Iceland
2        12348         Finland
3        12349           Italy
4        12350          Norway

In [16]:
customer_frm = pd.merge(frm, customer, how='inner')
customer_frm.head()

customer_id  total_order  total_day  avg_recency  max_recency  min_recency  \
0        12347          6.0        6.0    60.833333           90           37   
1        12348          3.0        3.0    94.333333          173           40   
2        12352         10.0        6.0    43.333333          182            5   
3        12356          2.0        2.0   151.500000          223           80   
4        12358          1.0        1.0   149.000000          149          149   

  frequency_rec  avg_total_order_monthly  max_total_order  min_total_order  \
0          high                     1.00                1                1   
1          high                     1.00                1                1   
2           low                     2.75                7                1   
3          high                     1.00                1                1   
4          high                     1.00                1                1   

   pricetotal  jumlah  basketsize         max     min        avg  \
0     4310.00       7  615.714286   35.657143  0.7200   3.383046   
1     1797.24       4  449.310000   60.000000  3.3000  14.493871   
2     2506.04       8  313.255000   47.062500  1.2375   3.685353   
3     2811.43       3  937.143333  108.000000  1.2500  15.883785   
4     1168.06       2  584.030000   80.000000  4.6800  30.738421   

  frequency_mrkt   country  
0            low   Iceland  
1           high   Finland  
2            low    Norway  
3           high  Portugal  
4           high   Austria

In [17]:
table_customer_frm = customer_frm[['customer_id','country','frequency_rec','frequency_mrkt',]]
table_customer_frm.head()

customer_id   country frequency_rec frequency_mrkt
0        12347   Iceland          high            low
1        12348   Finland          high           high
2        12352    Norway           low            low
3        12356  Portugal          high           high
4        12358   Austria          high           high

In [18]:
compare_table_customer_frm = table_customer_frm[(table_customer_frm['frequency_rec']=='low')&(table_customer_frm['frequency_mrkt']=='high')]
compare_table_customer_frm.head()

customer_id      country frequency_rec frequency_mrkt
11        12371  Switzerland           low           high
14        12377  Switzerland           low           high
15        12379      Belgium           low           high
19        12384  Switzerland           low           high
20        12386    Australia           low           high

From the analysis above, we get insight that even though they (customer) weren't quite often buying but once they bought, they bought a lot.

In [19]:
#Analyse which country is the customer coming from
compare_table_customer_frm['country'].value_counts().head()

United Kingdom    433
Germany            18
France             11
Switzerland         8
Portugal            5
Name: country, dtype: int64

There are 433 customers who are United Kingdom citizen.